backend/api/src/db/jig/player.rs

```diff
@@ -89,6 +91,107 @@ values ($1, $2, $3, $4, $5, $6)
     }
 }
 
+/// Creates new jig player session for a player
+// TODO: Fixme
+// TODO: Add Settings?
+pub async fn create_user_session(
+    db: &PgPool,
+    session_index: i16,
+    ip_addr: IPAddress,
+) -> Result<String, error::JigCode> {
+    let mut txn = db.begin().await?;
+
+    let jig_id = sqlx::query!(
+        //language=SQL
+        r#"
+        select jig_id
+        from jig_player_session
+        where index=$1
+        "#,
+        session_index
+    )
+    .fetch_optional(&mut txn)
+    .await?
+    .ok_or(error::JigCode::ResourceNotFound)?;
+
+    // get the user agent header and ip address
+    let ip_address = ip_addr.ip_addr.ok_or(error::JigCode::ResourceNotFound)?;
+    let user_agent = ip_addr.user_agent.ok_or(error::JigCode::ResourceNotFound)?;
+
+    // insert into the jig_player_session_instance table returning the instance_id
+    let jig_player_session_instance_id = sqlx::query!(
+        //language=SQL
+        r#"
+        insert into jig_player_session_instance (index, jig_id, ip_addr, user_agent)
+        values ($1, $2, $3, $4)
+        returning instance_id as "id: Uuid"
+        "#,
+        session_index,
+        jig_id.jig_id,
+        ip_address,
+        user_agent
+    )
+    .fetch_one(&mut txn)
+    .await?;
+
+    txn.commit().await?;
+
+    Ok(jig_player_session_instance_id.id.to_string())
+}
+
+/// Completes a jig player session for a player and updates play count
+pub async fn complete_user_session(
+    db: &PgPool,
+    jig_id: JigId,
+    ip_addr: IPAddress,
+    session_id: &str,
+) -> Result<(), error::JigCode> {
+    let txn = db.begin().await?;
+
+    let instance_id = Uuid::parse_str(&session_id).unwrap();
+
+    let jig_session_index_query = sqlx::query!(
+        //language=SQL
+        r#"
+        select *
+        from jig_player_session_instance
+        where instance_id = $1
+        "#,
+        instance_id,
+    )
+    .fetch_optional(db)
+    .await?
+    .ok_or(error::JigCode::ResourceNotFound)?;
+
+    // get the user agent header and ip address and check validity
+    let ip_address = ip_addr.ip_addr.ok_or(error::JigCode::ResourceNotFound)?;
+    let user_agent = ip_addr.user_agent.ok_or(error::JigCode::ResourceNotFound)?;
+
+    // unwrap because invalid ip/user agents are checked before putting them into the db
+    // confirms user agent and ip address match
+    if (jig_session_index_query.user_agent.unwrap()).ne(&user_agent)
+        | (jig_session_index_query.ip_addr.unwrap()).ne(&ip_address)
+    {
+        return Err(error::JigCode::ResourceNotFound);
+    }
+
```

Comment:
If the session instance is successfully completed, we should remove the row from the database